common/unsafe/src/main/java/org/apache/spark/unsafe/types/CalendarInterval.java

```diff
@@ -78,6 +81,27 @@ public int hashCode() {
     return Objects.hash(months, days, microseconds);
   }
 
+  @Override
+  public int compare(CalendarInterval that) {
+    long thisAdjustDays = this.microseconds / MICROS_PER_DAY + this.days + this.months * 30;
+    long thatAdjustDays = that.microseconds / MICROS_PER_DAY + that.days + that.months * 30;
+    long daysDiff = thisAdjustDays - thatAdjustDays;
+    if (daysDiff == 0) {
+      long msDiff = (this.microseconds % MICROS_PER_DAY) - (that.microseconds % MICROS_PER_DAY);
+      if (msDiff == 0) {
+        return 0;
+      } else if (msDiff > 0) {
+        return 1;
+      } else {
+        return -1;
+      }
+    } else if (daysDiff > 0){
```

Comment:
we can just return `daysDiff` if it's non-zero. The `Ordered` doesn't force us to return 1 or -1.

Reply:
the adjusted days might be long here

Reply:
ah i see